### PR TITLE
Refactor node controllers

### PR DESCRIPTION
**Pull request by andriisoldatenko**
## Description

Please include the following:

This PR implements some housekeeping work around node controller, including: 
- remove unused code 
- remove nodeDeletionPredicate from the event filter and handle this case properly in the Reconcile loop
- add more tests
- migrate from `time.Now` to `timeprovider`. 

## How can this be tested?

1) Deploy the operator and install dynakube. Then, delete the node and make sure we send a mark for the termination request.
2) Drain/Cordon node and wait ~1h and check logs.

## Checklist

- [x] Unit tests have been updated/added
- [x] PR is labeled accordingly with a single label
- [x] I have read and understood the [contribution guidelines](https://github.com/Dynatrace/dynatrace-operator/blob/main/CONTRIBUTING.md)

### CODE REVIEW

File: pkg/controllers/nodes/nodes_controller.go [hunk 1]
```diff
@@ -90,8 +78,8 @@ func (controller *Controller) Reconcile(ctx context.Context, request reconcile.R
 	var node corev1.Node
 	if err := controller.apiReader.Get(ctx, client.ObjectKey{Name: nodeName}, &node); err != nil {
 		if k8serrors.IsNotFound(err) {
-			log.Info("node was not found in cluster", "node", nodeName)
-			return reconcile.Result{}, nil
+			// if there is no node it means it get deleted or cordon or evicted
```
Comment (StefanHauth) [inline pkg/controllers/nodes/nodes_controller.go:81]:
```suggestion
			// if there is no node it means it get deleted, cordoned or evicted
```

**Review by StefanHauth — CHANGES_REQUESTED** [review]

File: pkg/controllers/nodes/nodes_controller.go [hunk 2]
```diff
@@ -228,19 +216,15 @@ func (controller *Controller) updateCache(ctx context.Context, nodeCache *Cache)
...
 	for _, cachedNodeName := range nodeCache.Keys() {
 		cachedNodeInCluster := false
 		for _, clusterNode := range nodeLst.Items {
 			if clusterNode.Name == cachedNodeName {
-				cachedNodeInfo, err := nodeCache.Get(cachedNodeName)
-				if err != nil {
-					log.Error(err, "failed to get node", "node", cachedNodeName)
-					return err
-				}
+				cachedNodeInfo, _ := nodeCache.Get(cachedNodeName)
```
[the remark this answers is not in the document]
Comment (andriisoldatenko) [reply pkg/controllers/nodes/nodes_controller.go:229]: added

File: pkg/controllers/nodes/nodes_controller.go [hunk 3]
```diff
@@ -90,8 +78,8 @@ func (controller *Controller) Reconcile(ctx context.Context, request reconcile.R
 	var node corev1.Node
 	if err := controller.apiReader.Get(ctx, client.ObjectKey{Name: nodeName}, &node); err != nil {
 		if k8serrors.IsNotFound(err) {
-			log.Info("node was not found in cluster", "node", nodeName)
-			return reconcile.Result{}, nil
+			// if there is no node it means it get deleted, cordoned or evicted
```
Comment (andriisoldatenko) [inline pkg/controllers/nodes/nodes_controller.go:81]:
```suggestion
			// if there is no node it means it get deleted
```

File: pkg/controllers/nodes/cache.go [hunk 4]
```diff
@@ -85,19 +87,10 @@ func (cache *Cache) Changed() bool {
 	return cache.Create || cache.upd
 }
 
-func (cache *Cache) ContainsKey(key string) bool {
```
Comment (andriisoldatenko) [inline pkg/controllers/nodes/cache.go:88]: dead code

**Review by 0sewa0 — APPROVED** [review]

**Pull request merged**